hail/src/main/scala/is/hail/expr/ir/IR.scala

```diff
@@ -170,31 +157,6 @@ final case class ApplyBinaryPrimOp(op: BinaryOp, l: IR, r: IR) extends IR
 final case class ApplyUnaryPrimOp(op: UnaryOp, x: IR) extends IR
 final case class ApplyComparisonOp(op: ComparisonOp[_], l: IR, r: IR) extends IR
 
-object MakeArray {
-  def unify(args: Seq[IR], typ: TArray = null): MakeArray = {
-    assert(typ != null || args.nonEmpty)
-    var t: TArray = typ
-    if (t == null) {
-      t = if (args.tail.forall(_.typ == args.head.typ)) {
-        TArray(args.head.typ)
-      } else TArray(args.head.typ.deepOptional())
-    }
-    assert(t.elementType.deepOptional() == t.elementType ||
-      args.forall(a => a.typ == t.elementType),
-      s"${ t.parsableString() }: ${ args.map(a => "\n    " + a.typ.parsableString()).mkString } ")
-
-    MakeArray(args.map { arg =>
-      if (arg.typ == t.elementType)
-        arg
-      else {
-        val upcast = PruneDeadFields.upcast(arg, t.elementType)
```

Comment:
this could cause the elementType to change (in a way other than requiredeness) previously, because t.elementType could be something entirely other than arg.elementType, and the assertion above on line 102 will never fall through to line 183 when typ is provided and has no requiredeness set (t.elementType.deepOptional() == t.elementType in that case).